source/extensions/common/wasm/null/null_vm.cc

```diff
@@ -17,6 +18,8 @@ namespace Common {
 namespace Wasm {
 namespace Null {
 
+std::atomic<int> active_vms_;
```

Comment:
static initialization fiasco?

Reply:
IMO the fact that "this is safe because the default constructor is empty and BSS is initialized to zero." seems like it's using expert knowledge to defend going against the grain of the style guide, which tries to simplify these things and make them easier to reason about without that expert knowledge.

IMO you should make a context object. If there's no place to establish it as a context available whenever you create VMs you could just make it a lazy-initialized singleton, using either the pattern in the style-guide or the simple Envoy wrapper class that captures that pattern.

I am also unclear on the semantics of your atomic as it looks to me like it's being accessed along with other atomics in a way that is functionally racy. So in this lazy-initialized objectI propose, you could put a mutex so you can make atomic the sequence that needs to be atomic. See below.


Also I didn't think we use trailing underscore for member variables. I don't remember what the right naming is for globals :) But if you use the lazy-init pattern you won't have a global variable, just a global function that accesses a local static.

Reply:
Not having expert knowledge and writing C++ is dangerous  :).

I have changed this to use MUTABLE_CONSTRUCT_ON_FIRST_USE which I think is simpler and clearer.  Lazy-initialization for the win.  PTAL.

Also, AFAICT Envoy uses a trailing underscore for all members.  I am looking at things like buffer_impl.h.

Reply:
nit; This is not a member variable; it's a global variable in a namespace.

Reply:
Yes, that is the point. It is a class static so that it will count the number of active VMs corresponding to that class.

Reply:
Changed to just use a Gauge.  PTAL